aptos-move/framework/aptos-framework/sources/Comparator.move

```diff
@@ -0,0 +1,146 @@
+/// Provides a framework for comparing two elements
+module AptosFramework::Comparator {
+    use Std::BCS;
+    use Std::Vector;
+
+    const EQUAL: u8 = 0;
+    const SMALLER: u8 = 1;
+    const GREATER: u8 = 2;
+
+    struct Result has drop {
+        inner: u8,
+    }
+
+    public fun is_equal(result: &Result): bool {
+        return result.inner == EQUAL
+    }
+
+    public fun is_smaller_than(result: &Result): bool {
+        return result.inner == SMALLER
+    }
+
+    public fun is_greater_than(result: &Result): bool {
+        result.inner == GREATER
+    }
+
+    // Performs a comparison of two types after BCS serialization.
+    public fun compare<T>(left: &T, right: &T): Result {
+        let left_bytes = BCS::to_bytes(left);
+        let right_bytes = BCS::to_bytes(right);
+
+        compare_u8_vector(left_bytes, right_bytes)
+    }
+
+    // Performs a comparison of two vector<u8>s or byte vectors
+    public fun compare_u8_vector(left: vector<u8>, right: vector<u8>): Result {
+        let left_length = Vector::length(&left);
+        let right_length = Vector::length(&right);
+
+        let idx = 0;
+
+        while (idx < left_length && idx < right_length) {
+            let left_byte = *Vector::borrow(&left, idx);
+            let right_byte = *Vector::borrow(&right, idx);
+
+            if (left_byte < right_byte) {
+                return Result { inner: SMALLER }
+            } else if (left_byte > right_byte) {
+                return Result { inner: GREATER }
+            };
+            idx = idx + 1;
+        };
+
+        if (left_length < right_length) {
+            Result { inner: SMALLER }
+        } else if (left_length > right_length) {
+            Result { inner: GREATER }
+        } else {
+            Result { inner: EQUAL }
+        }
+    }
+
+    #[test]
+    public fun test_strings() {
+        use Std::ASCII;
+
+        let value0 = ASCII::string(b"alpha");
+        let value1 = ASCII::string(b"beta");
+        let value2 = ASCII::string(b"betaa");
+
+        assert!(is_equal(&compare(&value0, &value0)), 0);
+        assert!(is_equal(&compare(&value1, &value1)), 1);
```

Comment:
nit: it made me wonder why there isn't `assert!(is_equal(&compare(&value1, &value1)), x);`

Reply:
I don't understand. Maybe you wanna see value2?

Reply:
😞